### ISSUE TITLE
ZIO 2 stable release

### ISSUE THREAD

**Issue by regiskuckaertz** [scalapb/zio-grpc#421]
Hi @thesamet 

We are keen to help you get the remaining work over the line for releasing a stable ZIO 2.0 version of the library. Would you mind listing the tickets that should be solved first? Or do you need help with the testing or something else?

**Comment by thesamet** [comment]
Thanks for offering help! The only things that I'm aware of that block a stable release are:
1. Enable .github/workflows/ci.yml for Scala 3 (I believe ther'es a compile error now)
2. There used to be a flaky test in e2eJVM. I am not sure if it's been fixed, but I wanted to ensure this isn't representing a bug. Need to run the suite multiple times to see if it shows up.

**Comment by regiskuckaertz** [comment]
Great! I fixed the compilation issue in #422 

Even though all the tests are passing (which should allow you to close #250 ) I see an exception being thrown in a spec that is not reported as a failure:
```
2022-08-14 18:14:54.700+0100 error [ManagedChannelOrphanWrapper] *~*~*~ Channel {0} was not shutdown properly!!! ~*~*~*
    Make sure to call shutdown()/shutdownNow() and wait until awaitTermination() returns true. 
java.lang.RuntimeException: ManagedChannel allocation site
  | => sat io.grpc.internal.ManagedChannelOrphanWrapper$ManagedChannelReference.<init>(ManagedChannelOrphanWrapper.java:93)
        at io.grpc.internal.ManagedChannelOrphanWrapper.<init>(ManagedChannelOrphanWrapper.java:53)
        at io.grpc.internal.ManagedChannelOrphanWrapper.<init>(ManagedChannelOrphanWrapper.java:44)
        at io.grpc.internal.ManagedChannelImplBuilder.build(ManagedChannelImplBuilder.java:630)
        at io.grpc.internal.AbstractManagedChannelImplBuilder.build(AbstractManagedChannelImplBuilder.java:297)
        at scalapb.zio_grpc.ClientCallsSpec$.unarySuite$$anonfun$1$$anonfun$1(ClientCallsSpec.scala:23)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:822)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:822)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:822)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:822)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:822)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:889)
        at zio.internal.FiberRuntime.runLoop(FiberRuntime.scala:936)
        at zio.internal.FiberRuntime.evaluateEffect(FiberRuntime.scala:324)
        at zio.internal.FiberRuntime.evaluateMessageWhileSuspended(FiberRuntime.scala:447)
        at zio.internal.FiberRuntime.drainQueueOnCurrentThread(FiberRuntime.scala:217)
        at zio.internal.FiberRuntime.run(FiberRuntime.scala:134)
```

do you have an idea off the top your head? Otherwise I'll dig deeper.

**Comment by thesamet** [comment]
It looks like there's no shutdown followed by awaitTermination on the client side similar to the change you recently made for the server side.

**Comment by regiskuckaertz** [comment]
I ran the benchmarks from https://github.com/scalapb/zio-grpc/issues/264 with the latest release and I was very surprised that the zio 2 version of zio-grpc was way slower than fs2-grpc. If anything I would have thought the runtime optimisation that was bandied about at the ZIO 2 release would have helped here, since there is a lot of `unsafeRun` everywhere, but it is definitely not the case. I will look into it in my own time.

**Comment by ashwinbhaskar** [comment]
@thesamet is there a plan to release zio 2 version of the library anytime soon? Just asking so that I can plan accordingly:)

**Comment by thesamet** [comment]
@ashwinbhaskar Currently there's a 0.6.0-test4 release. Help is still needed in debugging a flaky test so we can have a stable release.

**Comment by regiskuckaertz** [comment]
@thesamet I think I fixed it [here](https://github.com/scalapb/zio-grpc/blob/master/core/src/main/scalajvm/scalapb/zio_grpc/ZChannel.scala#L53) after all or is there another one you were thinking about? I ran the suite many times with no failure in sight.

**Comment by thesamet** [comment]
> @thesamet I think I fixed it [here](https://github.com/scalapb/zio-grpc/blob/master/core/src/main/scalajvm/scalapb/zio_grpc/ZChannel.scala#L53) after all or is there another one you were thinking about? I ran the suite many times with no failure in sight.

This is great! I also haven't seen this flaky test in the past few CI runs today.

**Comment by thesamet** [comment]
I spoke too soon: https://github.com/scalapb/zio-grpc/actions/runs/3456300753/jobs/5768932356#step:6:147

**Comment by regiskuckaertz** [comment]
Ah, right, here's the fix: https://github.com/scalapb/zio-grpc/pull/451

**Comment by lxwbr** [comment]
@thesamet do you think the latest 0.6.0 test release is production-ready or would you advise waiting until all flaky tests are fixed?

**Comment by lxwbr** [comment]
@thesamet thanks, yes we do use streaming API, although realistically I am not expecting cases in production where we would run into back-pressure situations. Will let you know if we should have any issues with the test release.

**Comment by ashwinbhaskar** [comment]
I tried out version `0.6.0-test8`. In my application I used zio version `2.0.0`. The src files compile successfully but tests fail with a very weird reason.
```
incompatible type in overriding
def bootstrap: zio.ZLayer[zio.ZIOAppArgs,Any,com.leadiq.dataiq.services.HelloWorldSpec.Environment] (defined in trait ZIOApp)
  with override val bootstrap: zio.ZLayer[zio.Scope,Any,zio.test.TestEnvironment] (defined in class ZIOSpecDefault);
 found   : zio.ZLayer[zio.Scope,Any,zio.test.TestEnvironment]
    (which expands to)  zio.ZLayer[zio.Scope,Any,zio.test.Annotations with zio.test.Live with zio.test.Sized with zio.test.TestConfig]
 required: zio.ZLayer[zio.ZIOAppArgs,Any,com.leadiq.dataiq.services.HelloWorldSpec.Environment]
    (which expands to)  zio.ZLayer[zio.ZIOAppArgs,Any,zio.test.Annotations with zio.test.Live with zio.test.Sized with zio.test.TestConfig]```
This is the test file 
```
import zio._
import zio.test._
import zio.test.Assertion._

import java.io.IOException

import HelloWorld._

object HelloWorld {
  def sayHello: ZIO[Any, IOException, Unit] =
    Console.printLine("Hello, World!")
}

object HelloWorldSpec extends ZIOSpecDefault {
  def spec = suite("HelloWorldSpec")(
    test("sayHello correctly displays output") {
      for {
        _      <- sayHello
        output <- TestConsole.output
      } yield assertTrue(output == Vector("Hello, World!\n"))
    }
  )
}

```
I asked to this on the zio discord channel. They said this could be something to do with dependency mismatch. I checked the source code and the `zio2` branch of this repo has zio version `2.0.0-RC1` while I am using `2.0.0`. Could this be the cause of the error?

**Comment by thesamet** [comment]
Hi @ashwinbhaskar , the latest release of zio-grpc is version 0.6.0-rc2. It is made out of the `master` branch and depends on zio `2.0.9`.

**Comment by ashwinbhaskar** [comment]
@thesamet thanks for the info! I will try out this version.

**Comment by thesamet** [comment]
Heads up that the last anticipated breaking change for the stable zio 2 is coming through #466 . I think this will be a welcome change as users expressed in the past that they prefer all their error types to be subtypes of `Throwable`. However, please speak up if you have concerns.

**Comment by regiskuckaertz** [comment]
🚀

**Comment by thesamet** [comment]
I am seeing a flaky test in our CI, if anyone has a bit of time to investigate, it would be really appreciated: https://github.com/scalapb/zio-grpc/issues/487

**Comment by omidb** [comment]
@thesamet Thanks for the amazing project. it would be great to see this merged and do another benchmark test with https://github.com/LesnyRumcajs/grpc_bench

**Comment by cipriansofronia** [comment]
Hi, not sure if this issue is the best place to raise this, but it's related to the ZIO2 release so I'm gonna share my findings here. Please advise if I have to move it to a new issue. 

We recently migrated to ZIO2, so zio-grpc got bumped to `0.6.0-rc5`. We've noticed some performance issues after the upgrade while performing streaming calls. Going to attach some screenshots of the IntelliJ Profiler. 

On the upper hand of the screenshot there is the timeline of the previous version of zio-grpc (`0.5.1`) running with ZIO1 and bellow we have the zio-grpc version `0.6.0-rc5` with ZIO2. 
The streaming request is kept alive for a few seconds for both versions but we can notice some differences between these two. On `0.5` there is a small spike when the request is triggered but then the CPU goes down and it stays like that, on `0.6` the CPU load is bigger and it says like that for the duration of the streaming request (not shown here in the screenshots but if we trigger more requests then the CPU goes even higher). 
There are no other changes in our service, only the migration to ZIO2 and bumping zio-grpc, so not doing anything extra on the new version for the same streaming call. The only thing different I noticed is that `0.6` is allocating more resources while performing `serverStreamingWithBackpressure` and underneath is calling ZIO2 internals. Using ZIO2 streaming so far in other parts of the app did not present performance issues.

<img width="3004" alt="Screenshot 2023-06-15 at 08 17 42" src="https://github.com/scalapb/zio-grpc/assets/9365666/d476e5a0-0d83-4495-8c67-f0201cdc6276">
<img width="3006" alt="Screenshot 2023-06-15 at 08 16 18" src="https://github.com/scalapb/zio-grpc/assets/9365666/375f6621-70ed-4d0d-8426-dac3c8a2dc18">



We isolated even more the streaming call, we created a separate project with zio-grpc `0.6` where we only do:
```scala
ZStream(<zio_grpc_generated_class>.defaultInstance)
      .repeat(Schedule.spaced(10.second))
```
So, nothing CPU consuming and we got the same behaviour. 

Please advise if more informations are needed.

**Comment by thesamet** [comment]
@cipriansofronia thanks for reporting, please start a new GitHub issue for the performance issue, so the discussion here can remain focused on release blockers at a higher level

**Comment by guizmaii** [comment]
+1

**Comment by thesamet** [comment]
I'm expecting final RC by this weekend, which would be followed by a final release before end of July.

**Comment by thesamet** [comment]
We have 0.6.0-rc6 released. Please give it a try - if no show stoppers are reported in the next two weeks, we'll call it final.

**Comment by thesamet** [comment]
Update: we're currently blocked from calling the latest RC stable due to the resurfaced issue #537. I would need someone to debug this, or alternatively, we will pass on backpressure for 0.6.0 and introduce this as an optional behavior after it stabilizes in a future release.

**Comment by regiskuckaertz** [comment]
That doesn’t seem related to backpressure or streaming, perhaps closing the scope in the `onExit` handler is wrong, I’ll take a look in a couple of days.

**Comment by ghostdogpr** [comment]
@regiskuckaertz yeah `scope.close` will trigger the interruption of the fiber but because you are in `onExit` you are already in a finalizer and inside an uninterruptible region so it hangs there? What was the reason for https://github.com/scalapb/zio-grpc/pull/515 ? It seems the fiber is about to terminate anyway when it reaches that point.

**Comment by ghostdogpr** [comment]
Actually I tried it and it seems to work 😄 I don't reproduce the issue at all.

**Comment by regiskuckaertz** [comment]
I think it's fine like but upon looking at the code for `forkDaemon` I wonder why there is a problem in the first place.

**Comment by regiskuckaertz** [comment]
@thesamet do you want to revert that change before the release?

**Comment by thesamet** [comment]
@regiskuckaertz sure - can you send a PR?

**Comment by Gregory-Berkman-Imprivata** [comment]
The bug reported in #537 is not actually the fault of zio-grpc and should not block the release. Hopefully this means the official release is no longer blocked?

**Comment by thesamet** [comment]
I've just released 0.6.0.
